isis/src/control/objs/CsmBundleObservation/CsmBundleObservation.cpp

```diff
@@ -0,0 +1,876 @@
+                                       QString instrumentId, BundleTargetBodyQsp bundleTargetBody) : AbstractBundleObservation(image, observationNumber, instrumentId, bundleTargetBody) {
+    if (bundleTargetBody) {
+      QString msg = "Target body parameters cannot be solved for with CSM observations.";
+      throw IException(IException::User, msg, _FILEINFO_);
+    }
+  }
+
+
+  /**
+   * Creates a copy of another CsmBundleObservation.
+   *
+   * @param src Reference to the CsmBundleObservation to copy
+   */
+  CsmBundleObservation::CsmBundleObservation(const CsmBundleObservation &src) : AbstractBundleObservation(src) {
+    m_solveSettings = src.m_solveSettings;
+    m_paramIndices = src.m_paramIndices;
+  }
+
+
+  /**
+   * Destructor.
+   *
+   * Contained BundleImages will remain until all shared pointers are deleted.
+   */
+  CsmBundleObservation::~CsmBundleObservation() {
+  }
+
+
+  /**
+   * Assignment operator
+   *
+   * Assigns the state of the source CsmBundleObservation to this CsmBundleObservation
+   *
+   * @param CsmBundleObservation Reference to the source CsmBundleObservation to assign from
+   *
+   * @return @b CsmBundleObservation& Reference to this CsmBundleObservation
+   */
+  CsmBundleObservation &CsmBundleObservation::operator=(const CsmBundleObservation &src) {
+    if (&src != this) {
+      m_solveSettings = src.m_solveSettings;
+      m_paramIndices = src.m_paramIndices;
+    }
+    return *this;
+  }
+
+
+  /**
+   * Set solve parameters
+   *
+   * @param solveSettings The solve settings to use
+   *
+   * @return @b bool Returns true if settings were successfully set
+   *
+   * @internal
+   *   @todo initParameterWeights() doesn't return false, so this methods always
+   *         returns true.
+   */
+  bool CsmBundleObservation::setSolveSettings(CsmBundleObservationSolveSettingsQsp solveSettings) {
+    // TODO implement for CSM
+
+    m_solveSettings = solveSettings;
+
+    CSMCamera *csmCamera = dynamic_cast<CSMCamera*>(front()->camera());
+
+    m_paramIndices = csmCamera->getParameterIndices(solveSettings->solveSet());
+    int nParams = m_paramIndices.size();
+
+    m_weights.resize(nParams);
+    m_weights.clear();
+    m_corrections.resize(nParams);
+    m_corrections.clear();
+    m_adjustedSigmas.resize(nParams);
+    m_adjustedSigmas.clear();
+    m_aprioriSigmas.resize(nParams);
+
+    for (int i = 0; i < nParams; i++) {
+      m_aprioriSigmas[i] = csmCamera->getParameterCovariance(m_paramIndices[i], m_paramIndices[i]);
+    }
+
+    return true;
+  }
+
+
+  // TODO change this to CSM solve settings once compute partials is updated
+  /**
+   * Accesses the solve settings
+   *
+   * @return @b const BundleObservationSolveSettingsQsp Returns a pointer to the solve
+   *                                                    settings for this CsmBundleObservation
+   */
+  const BundleObservationSolveSettingsQsp CsmBundleObservation::solveSettings() {
```

Comment:
This returns the wrong type of solve settings because of how the abstract interface works. It expects to return an ISIS settings. The only place it's used is in computePartials, so once we get that updated we should remove this from the abstract class so that we can have unique settings.